src/main/java/org/breedinginsight/brapps/importer/services/FileImportService.java

```diff
@@ -0,0 +1,285 @@
+    }
+
+    public ImportMapping updateMapping(UUID programId, AuthenticatedUser actingUser, UUID mappingId,
+                                       ImportMapping mappingRequest, Boolean validate) throws
+            DoesNotExistException, AuthorizationException, HttpStatusException, UnprocessableEntityException {
+
+        if (!programService.exists(programId))
+        {
+            throw new DoesNotExistException("Program id does not exist");
+        }
+
+        if (!programUserService.existsAndActive(programId, actingUser.getId())) {
+            throw new AuthorizationException("User not in program");
+        }
+
+        Optional<ImportMapping> optionalImportMapping = importMappingDAO.getMapping(mappingId);
+        if (optionalImportMapping.isEmpty()) {
+            throw new DoesNotExistException("Mapping with that id does not exist");
+        }
+        ImportMapping importMapping = optionalImportMapping.get();
+
+        // If validate is true, validate the mapping
+        if (validate) {
+            mappingManager.map(mappingRequest, importMapping.getFile());
+        }
+
+        // Save the mapping
+        String json = null;
+        try {
+            json = objectMapper.writeValueAsString(mappingRequest.getMapping());
+        } catch(JsonProcessingException e) {
+            log.error("Problem converting traits json", e);
+            // If we didn't catch this in parsing to the class, this is a server error
+            throw new InternalServerException("Problem converting mapping json", e);
+        }
+
+        ImportMappingEntity importMappingEntity = importMappingDAO.fetchOneById(mappingId);
+        importMappingEntity.setName(mappingRequest.getName());
+        importMappingEntity.setImportTypeId(mappingRequest.getImportTypeId());
+        importMappingEntity.setMapping(JSONB.valueOf(json));
+        importMappingEntity.setDraft(mappingRequest.getDraft());
+        importMappingDAO.update(importMappingEntity);
+
+        Optional<ImportMapping> updatedMapping = importMappingDAO.getMapping(importMappingEntity.getId());
+
+        return updatedMapping.get();
+    }
+
+    public Boolean exists(UUID mappingId) {
+        return importMappingDAO.existsById(mappingId);
+    }
+
+    public ImportPreviewResponse uploadData(UUID programId, UUID mappingId, AuthenticatedUser actingUser, CompletedFileUpload file, Boolean commit)
+            throws DoesNotExistException, AuthorizationException, UnsupportedTypeException, HttpStatusException, UnprocessableEntityException {
+
+        Optional<Program> optionalProgram = programService.getById(programId);
+        if (!optionalProgram.isPresent())
+        {
+            throw new DoesNotExistException("Program id does not exist");
+        }
+        Program program = optionalProgram.get();
+
+        if (!programUserService.existsAndActive(programId, actingUser.getId())) {
+            throw new AuthorizationException("User not in program");
+        }
```

Comment:
This code could be made into a reusable method

Reply:
pushed